keps/sig-node/20190212-ephemeral-containers.md

```diff
@@ -358,29 +365,119 @@ There are no limits on the number of Ephemeral Containers that can be created in
+method in the API to restart an Ephemeral Container.  Creators of Ephemeral
+Containers are expected to choose a new, unused name.
+
+### Configurable Security Policy
+
+The ability to add containers to a pod implies security trade offs. We've
+received the following requirements and feedback on the alpha implementation:
+
+1.  Admission controllers should be able to enforce policy based on the
+    cumulative pod specification, so operations that prune information,
+    such as removing Ephemeral Containers should not be allowed.
+1.  Restarting a pod is disruptive, so for reasons of operation, security,
+    and resource hygiene it should be possible to delete Ephemeral Containers
+    via the API. ([#84764]).
+1.  Ephemeral Containers could allow privilege escalation greater than that
+    of the initial pod, so setting a custom security context should not be
+    allowed.
+1.  Ephemeral Containers, which are initiated by humans for debugging purposes,
+    should be allowed a more permissive security context regular containers.
+    ([#53188])
+
+Rather than choosing a security posture, policy for Ephemeral Containers
+will be configurable by cluster administrators via 
+[`PodSecurityPolicy`][psp-docs].
+
+[#53188]: https://issues.k8s.io/53188
+[#84764]: https://issues.k8s.io/84764#issuecomment-602004629
+[psp-docs]: https://kubernetes.io/docs/concepts/policy/pod-security-policy/
+
+#### Removing Ephemeral Containers
+
+- Default when `PodSecurityPolicy` controller enabled: disallowed
+- Default when `PodSecurityPolicy` controller disabled: allowed
+
+Ephemeral Containers will stop when their command exits, such as exiting a
+shell, and they will not be restarted.  Unlike `kubectl exec`, processes in
+Ephemeral Containers will not receive an EOF if their connection is
+interrupted. Without the ability to remove an Ephemeral Container via the
+API, the only way to exit the container is to send it an OS signal.
+
+Removing an Ephemeral Container from the list of Ephemeral Containers in the
+pod spec will result in the kubelet killing the container. The kubelet will
+cease to publish a `ContainerStatus` for this container.
+
+When [`PodSecurityPolicy`][psp-docs] is enabled, removing Ephemeral Containers
+is disabled by default, but it may be enabled by setting a new field
+`allowEphemeralContainerDeletion` in the security policy.
+
+#### Specifying Security Context
+
+- Default when `PodSecurityPolicy` controller enabled: allowed
+- Default when `PodSecurityPolicy` controller disabled: allowed
+
+`PodSecurityPolicy` already contains several fields allowing the restriction of
+fields within a container's security context, and it allows binding different
+policies to different roles. These existing options should be sufficient for
+cluster administrators to restrict security context as the security policy
+controller also checks Ephemeral Containers, including during updates.
```

Comment:
This works for PSP since the policy can be bound to the user. For most other policy controllers, the policy is bound to the pod through labels or namespaces, so it would be harder to have a policy enforced on a namespace but grant exceptions to users authorized users.

One approach is to have separate policies for regular containers and ephemeral containers, and rely on RBAC to limit ephemeral container access. I think this would work as long as operations on ephemeral containers are required to go through a subresource, and cannot be done as part of a regular pod update request. Is that the case?

Reply:
Having separate policies for ephemeral containers works for me. I think it'll be easier to understand. Is there an existing policy controller where this would fit?

That's correct, ephemeral containers may only be modified via the `/ephemeralcontainers` subresource.